---
title: "Results"
author: "Jakob"
output: 
    html_document:
      code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```



Setup
=====

Load packages
```{r}
library(tidyverse)
library(drake)
library(here)
library(rtracklayer)
library(rptR)
library(inovaDNMs)
library(inovaDNMs102)
library(dnmVarianceComponents)
library(ggbeeswarm)
library(foreach)
library(poissonMutSim)
library(car)

set.seed(4321)
```

Get data
--------

Get the DNMs & have them in the right format
```{r}
inova1 <- 
  inovaDNMs::obs %>%
  dplyr::rename(
    SampleID = internalID,
    Variant = Variant, 
    Chr = Chromosome,
    Position = Start.position)  %>%
  full_join(inovaDNMs::trios %>% 
              filter(in.data), 
            by=c(SampleID = "Trio.ID")) %>%
  dplyr::rename(familyNr = Family.Study.ID) %>%
  mutate(familyNr = if_else(familyNr %in% c("101-112", "101-689"), "101-112/101-689", familyNr)) %>% # properly label sibling pair (see Wendy's mail 2nd oct 2016)
  mutate(PROBANDID=paste0(SampleID, "-03")) %>%
  left_join(inovaDNMs::seqReactions[,c("software", "name")] %>%
              mutate(batch = substr(software, 1,6)),
            by=c(Proband.Study.ID="name"))

inova2 <- 
  inovaDNMs102::allDNMs %>%
  dplyr::rename(
    SampleID = sampleName,
    Variant = alt, 
    Chr = chrom,
    Position = start) %>%
  full_join(inovaDNMs102::trios, 
            by=c(SampleID = "PROBANDID")) %>%
  dplyr::rename(familyNr = FAMILYID) %>%
  mutate(end = NULL) %>% 
  left_join(inovaDNMs102::siblings %>% 
              filter(!SiblingPairId == "Fam12" & !SiblingPairId == "Sib14") %>% #clean incomplete families
              dplyr::select(-PROBANDID) %>% #drop double entries
              distinct(), 
            by=c("familyNr"="FAMILYID")) %>%
  mutate(familyNr = if_else(mode=="siblings" & !is.na(mode), SiblingPairId, familyNr)) %>% 
  left_join(inovaDNMs102::qcStats %>%
              dplyr::select(pipeline, STUDY_ID) %>%
              mutate(STUDY_ID=toupper(STUDY_ID), batch=pipeline),
            by=c(SampleID="STUDY_ID"))

#sasani DNMs
sasani <- 
  read_delim("https://github.com/quinlan-lab/ceph-dnm-manuscript/raw/master/data/second_gen.dnms.txt",
             delim = "\t",
             col_types = cols(chrom = col_character(),
                              new_family_id = col_character())) %>%
  bind_rows(read_delim("https://github.com/quinlan-lab/ceph-dnm-manuscript/raw/master/data/third_gen.dnms.txt",
                       delim = "\t",
                       col_types = cols(chrom = col_character(),
                                        new_family_id = col_character()))) %>%
  bind_rows(read_delim("https://github.com/quinlan-lab/ceph-dnm-manuscript/raw/master/data/post-pgcs.dnms.txt",
                       delim = "\t",
                       col_types = cols(chrom = col_character(),
                                        new_family_id = col_character()))) %>%
  bind_rows(read_delim("https://github.com/quinlan-lab/ceph-dnm-manuscript/raw/master/data/gonosomal.dnms.txt",
                       delim = "\t",
                       col_types = cols(chrom = col_character(),
                                        new_family_id = col_character()))) %>%
  mutate(Chr = paste0("chr", chrom)) %>%
  dplyr::rename(
    SampleID = new_sample_id,
    Variant = ref, 
    Position = start,
    fathersAgeAtConceptionInYears = paternal_age_at_birth,
    mothersAgeAtConceptionInYears = maternal_age_at_birth,
    familyNr = new_family_id
  )

autism1 <- dnmVarianceComponents::dnmDataAut
autism2 <- dnmVarianceComponents::dnmDataAut2

```


Prepare the genomic regions to filter DNMs
```{r}
stopifnot(file.exists(here("data/master_repeat_file.bed.gz")))
stopifnot(file.exists(here("data/recent_repeat_file.bed.gz")))

uniqGenome <- 
  import(here("data/master_repeat_file.bed.gz")) %>% 
  gaps()
seqlevelsStyle(uniqGenome) <- "UCSC"
recRepGenome <- 
  import(here("data/recent_repeat_file.bed.gz"))
seqlevelsStyle(recRepGenome) <- "UCSC"
ancientRepGenome <- setdiff(gaps(uniqGenome), recRepGenome)
allGenome <- SeqinfoForUCSCGenome("hg19") %>% as("GRanges")

```


Define functions
----------------

For filtering DNMs
```{r}
filterMuts <- 
  function(muts, ## needs columns "SampleID", Variant, Chr, Position
           filter){
    muts %>% 
      makeGRangesFromDataFrame(seqnames.field = "Chr", 
                               start.field = "Position", 
                               end.field = "Position", 
                               ignore.strand = TRUE, 
                               keep.extra.columns = TRUE) %>% 
      subsetByOverlaps(filter) %>%
      as.data.frame() %>%
      as.tibble() %>%
      mutate(Chr=seqnames, Position=start)
  }

filterAndCalculate <- function(sasani, uniqGenome, withBatch = FALSE) {
  if(withBatch) {
    filterMuts(sasani, uniqGenome) %>% 
    group_by(SampleID, 
             fathersAgeAtConceptionInYears, 
             mothersAgeAtConceptionInYears, 
             familyNr,
             batch) %>% 
      count() %>% 
      calcRelativeVarCorsWithCi(withBatch = TRUE)
  } else {
    filterMuts(sasani, uniqGenome) %>% 
      group_by(SampleID, 
               fathersAgeAtConceptionInYears, 
               mothersAgeAtConceptionInYears, 
               familyNr) %>% 
      count() %>% 
      calcRelativeVarCorsWithCi(withBatch = FALSE)
  }
}
```



Twins versus PAMUCs
===================

Function for formatting twin versus PAMUC data
```{r}
getTwins <- function(dnmData101) {
  dnmData101 %>% 
    add_count(familyNr) %>% 
    filter(n==2) %>% 
    group_by(familyNr) %>% 
    do(mutate(., 
              twins = (diff(fathersAgeAtConceptionInYears)==0) & (diff(mothersAgeAtConceptionInYears)==0))) %>% 
    filter(twins) %>%
    filter(familyNr != "101-306") %>% #weirdly annotated in 101 set
    ungroup() %>%
    dplyr::select(-twins)
}

getPamucs <- function(dnmData101, maxAgeDiff) {
  twinFamilies <- 
    dnmData101 %>% 
    getTwins() %>%
    pull(familyNr)
  
  dnmData101 %>%
    filter(!(familyNr %in% twinFamilies)) %>%
    arrange(fathersAgeAtConceptionInYears, mothersAgeAtConceptionInYears) %>% 
    mutate(ageClusterNumber=
             as.character(
               cutree(
                 hclust(dist(cbind(fathersAgeAtConceptionInYears, mothersAgeAtConceptionInYears), method = "manhattan")), 
                 h=maxAgeDiff/365)))  %>% 
    add_count(ageClusterNumber) %>%
    filter(n==2)
}

assembleTwinsVsUnrelated <- 
  function(dnmData101, maxAgeDiff) {
    twins <- 
      getTwins(dnmData101) %>% 
      group_by(familyNr) %>%
      nest() %>%
      mutate(no.pairs=map_int(data, ~nrow(.))) %>%
      filter(no.pairs==2) %>%
      mutate(diff_n=map_int(data, ~max(abs(diff(.$no.DNMs)))),
             set = "dizygote twins") %>% 
      mutate(group = as.character(familyNr)) %>%
      select(-familyNr)
    
    pamucs <- 
      getPamucs(dnmData101, maxAgeDiff) %>% 
      group_by(ageClusterNumber) %>%
      nest() %>%
      mutate(no.pairs=map_int(data, ~nrow(.))) %>%
      filter(no.pairs==2) %>%
      mutate(diff_n=map_int(data, ~max(abs(diff(.$no.DNMs)))),
             set = "PAMUCs") %>% 
      mutate(group = map_chr(data, ~do.call(paste, as.list(.$familyNr)))) %>%
      select(-ageClusterNumber)
    
    bind_rows(
      twins,
      pamucs
    )
  }
```

Getting main data like in thesis version, except for applying filters for unique genome regions:
```{r}
maxAgeDiff <- 43

twinsVsUnrelated <-
  rbind(
    assembleTwinsVsUnrelated(inova1 %>% 
                               filterMuts(uniqGenome) %>%
                               group_by(SampleID, 
                                        fathersAgeAtConceptionInYears, 
                                        mothersAgeAtConceptionInYears, 
                                        familyNr) %>%
                               count() %>% 
                               dplyr::rename(no.DNMs = n) %>%
                               ungroup(), 
                             maxAgeDiff) %>% 
      mutate(cohort="inova1"), 
    assembleTwinsVsUnrelated(inova2 %>% 
                               filterMuts(uniqGenome) %>% 
                               group_by(SampleID, 
                                        fathersAgeAtConceptionInYears, 
                                        mothersAgeAtConceptionInYears, 
                                        familyNr) %>%
                               count() %>% 
                               dplyr::rename(no.DNMs = n) %>%
                               ungroup(), maxAgeDiff) %>% 
      mutate(cohort="inova2"),
    assembleTwinsVsUnrelated(autism1 %>% mutate(no.DNMs=unique_snv), maxAgeDiff) %>% 
      mutate(cohort="autism1"),
    assembleTwinsVsUnrelated(autism2 %>% mutate(no.DNMs=snvs_unique), maxAgeDiff) %>% 
      mutate(cohort="autism2")
    ) %>%
  mutate(fatherAge=map_dbl(data, ~mean(.$fathersAgeAtConceptionInYears)),
         cohort=fct_inorder(cohort))
```

Visualization of the data for all cohorts apart
```{r}
twinsVsUnrelated %>% 
  filter(set=="dizygote twins") %>% 
  mutate(fatherAge=map(data, ~.$fathersAgeAtConceptionInYears), no.DNMs=map(data, ~.$no.DNMs), group=seq_along(no.pairs)) %>% 
  dplyr::select(-data) %>% 
  unnest() %>% 
  ggplot(aes(x=fatherAge, y=no.DNMs, group=group)) + 
  geom_point(col="#008000") + 
  geom_path(col="#008000") +
  labs(x="Father's age at conception in years",
       y="Dizygotic twins' no. DNMs") + 
  facet_wrap(~cohort) + 
  ggpubr::theme_pubr()

twinsVsUnrelated %>% 
  filter(set=="PAMUCs") %>% 
  mutate(fatherAge=map(data, ~.$fathersAgeAtConceptionInYears), no.DNMs=map(data, ~.$no.DNMs), group=seq_along(no.pairs)) %>% 
  dplyr::select(-data) %>% 
  unnest() %>% 
  ggplot(aes(x=fatherAge, y=no.DNMs, group=group)) + 
  geom_point(col="#2b2b2b") + 
  geom_path(col="#2b2b2b") +
  labs(x="Father's age at conception in years",
       y="PAMUCs' no. DNMs") + 
  facet_wrap(~cohort) + 
  ggpubr::theme_pubr()


twinsVsUnrelated %>%
  ggplot(aes(x=cohort, y=diff_n, col=set)) + 
  geom_violin(aes(fill=set)) + 
  geom_boxplot(width=0.9)+
  scale_color_manual(values = c("PAMUCs"="#2b2b2b", "dizygote twins"="#008000"), 
                     labels = c("PAMUCs"="PAMUCs", "dizygote twins"="Dizygotic twins")) + 
    scale_fill_manual(values = c("PAMUCs"="#2b2b2b", "dizygote twins"="#008000"), 
                     labels = c("PAMUCs"="PAMUCs", "dizygote twins"="Dizygotic twins")) + 
  labs(x="", y="Difference in no. DNMs",
       col="Set") + 
  ggpubr::theme_pubr()

```

Numbers
-------

```{r}
twinsVsUnrelated %>% 
  group_by(set, cohort) %>% 
  summarize(medDiff=median(diff_n), 
            mx=max(diff_n), 
            mn=min(diff_n))
twinsVsUnrelated %>% 
  group_by(set, cohort) %>% 
  count()
```

Tests
-----

First, are twin differences related to paternal age?
```{r}
# linear regression
twinsVsUnrelated %>%
  filter(set=="dizygote twins") %>%
  dplyr::select(-data) %>% 
  with(lm(diff_n~fatherAge)) %>% 
  broom::tidy() %>% 
  filter(term=="fatherAge") %>% 
  pull(p.value)

# heteroscedacity
twinsVsUnrelated %>%
  mutate(fatherAge=map_dbl(data, ~mean(.$fathersAgeAtConceptionInYears))) %>%
  filter(set=="dizygote twins") %>%
  dplyr::select(-data) %>%
  lm(diff_n~fatherAge, data=.) %>%
  lmtest::bptest() %>% 
  with(.$p.value)
```

Second, are the twin differences of PAMUCs larger than of twins?
```{r}
twinsVsUnrelated %>% 
  filter(cohort != "sasani") %>%
  group_by(cohort) %>% 
  do(with(.,
          broom::tidy(wilcox.test(diff_n~set, 
                                  alternative="less", 
                                  exact=FALSE))))    # use a normal approximation, as there are ties in the data
twinsVsUnrelated %>% 
  filter(!(cohort %in% c("sasani", "autism1"))) %>% 
  with(.,broom::tidy(wilcox.test(diff_n~set, 
                                 alternative="less", 
                                 exact=FALSE)))
```

Third, are PAMUC differences related to paternal age?
```{r}
#linear regression
twinsVsUnrelated %>%
  mutate(fatherAge=map_dbl(data, ~mean(.$fathersAgeAtConceptionInYears))) %>% 
  filter(set!="dizygote twins") %>%
  dplyr::select(-data) %>% 
  with(lm(diff_n~fatherAge)) %>% 
  broom::tidy() %>% 
  filter(term=="fatherAge") %>% 
  pull(p.value)

#heteroscedacity
twinsVsUnrelated %>%
  mutate(fatherAge=map_dbl(data, ~mean(.$fathersAgeAtConceptionInYears))) %>%
  filter(set!="dizygote twins") %>%
  dplyr::select(-data) %>%
  lm(diff_n~fatherAge, data=.) %>%
  lmtest::bptest() %>% 
  with(.$p.value)

```


Residual analysis
=================

```{r}
compareFamilyResidualDeltas <- function(autism1, n=1000) {
  simpMod <- lm(n ~ fathersAgeAtConceptionInYears + mothersAgeAtConceptionInYears, autism1)
  obs <- 
    autism1 %>%
    mutate(resi = residuals(simpMod)) %>%
    group_by(familyNr) %>% 
    nest() %>% 
    mutate(no.offspring = map_dbl(.$data, nrow)) %>% 
    filter(no.offspring == 2) %>% 
    mutate(delta = map_dbl(.$data, ~abs(diff(.$resi)))) %>%
    dplyr::select(-data, -no.offspring) %>%
    mutate(set="observed")
  sim <- 
    foreach(1:n,
            .combine = bind_rows,
            .packages = c("tidyverse")) %dopar% {
              autism1 %>%
                mutate(resi = sample(residuals(simpMod))) %>%
                group_by(familyNr) %>% 
                nest() %>% 
                mutate(no.offspring = map_dbl(.$data, nrow)) %>% 
                filter(no.offspring == 2) %>% 
                mutate(delta = map_dbl(.$data, ~abs(diff(.$resi)))) %>%
                dplyr::select(-data, -no.offspring) %>%
                mutate(set="simulated")
            }
  simObs <- bind_rows(obs, sim)
}
```


```{r}
doParallel::registerDoParallel(cores=7)

set.seed(4321)

filterMuts(inova1, uniqGenome) %>% 
  group_by(SampleID, 
           fathersAgeAtConceptionInYears, 
           mothersAgeAtConceptionInYears, 
           familyNr) %>% 
  dplyr::count() %>%
  ungroup() %>% 
  compareFamilyResidualDeltas()   %>%
  with(wilcox.test(delta ~ set))

filterMuts(inova2, uniqGenome) %>% 
  group_by(SampleID, 
           fathersAgeAtConceptionInYears, 
           mothersAgeAtConceptionInYears, 
           familyNr) %>% 
  dplyr::count() %>%
  ungroup() %>% 
  compareFamilyResidualDeltas()  %>%
  with(wilcox.test(delta ~ set))

compareFamilyResidualDeltas(autism1 %>% mutate(n=unique_snv)) %>%
  with(wilcox.test(delta ~ set))
compareFamilyResidualDeltas(autism2 %>% mutate(n=snvs_unique))  %>%
  with(wilcox.test(delta ~ set))

```

```{r}
agg_res <-
  bind_rows(
    compareFamilyResidualDeltas(autism1 %>% mutate(n=unique_snv) %>% mutate(familyNr = as.character(familyNr))) %>%
      mutate(dataset = "autism1"),
    compareFamilyResidualDeltas(autism2 %>% mutate(n=snvs_unique) %>% mutate(familyNr = as.character(familyNr))) %>%
      mutate(dataset = "autism2"),
    filterMuts(inova1, uniqGenome) %>% #0.41
      group_by(SampleID, 
               fathersAgeAtConceptionInYears, 
               mothersAgeAtConceptionInYears, 
               familyNr) %>% 
      dplyr::count() %>%
      ungroup() %>% 
      compareFamilyResidualDeltas() %>%
      mutate(dataset = "inova1"),
    filterMuts(inova2, uniqGenome) %>% 
      group_by(SampleID, 
               fathersAgeAtConceptionInYears, 
               mothersAgeAtConceptionInYears, 
               familyNr) %>% 
      dplyr::count() %>%
      ungroup() %>% 
      compareFamilyResidualDeltas() %>%
      mutate(dataset = "inova2")
  ) %>% 
  mutate(dataset = fct_inorder(dataset))

ggplot(agg_res, aes(x=dataset, y=delta, group=paste(dataset, set))) +
  geom_violin(aes(fill=set)) + 
  geom_boxplot(fill=NA, width=0.9) + 
  ggpubr::theme_pubr() + 
  labs(x=NULL, y="Parental age-corrected difference in no. DNMs")

agg_res %>% 
  with(wilcox.test(delta ~ set))
```


Variance component estimations
==============================

Function for running the variance component estimations
```{r}
calcRelativeVarCorsWithCi <- function(dnms, 
                                      withBatch=TRUE,
                                      nsim = 500) {
  doublecheckFitConvergence(dnms, withBatch) # assert that we can ignore model convergence warnings
  if(withBatch) {
    estim <-
      rpt(n ~ fathersAgeAtConceptionInYears + mothersAgeAtConceptionInYears + (1|familyNr) + (1|batch),
          data = dnms,
          grname=c("familyNr", "batch", "Fixed", "Residual"),
          nboot=nsim,
          npermut = 0,
          adjusted = FALSE,
          datatype = "Gaussian",
          parallel = TRUE)
  } else {
    estim <-
      rpt(n ~ fathersAgeAtConceptionInYears + mothersAgeAtConceptionInYears + (1|familyNr),
          data = dnms,
          grname=c("familyNr", "Fixed", "Residual"),
          nboot=nsim,
          npermut = 0,
          adjusted = FALSE,
          datatype = "Gaussian",
          parallel = TRUE)
  }
  res <- 
    bind_cols(estimate=as.double(estim$R), 
              estim$CI_emp %>% rownames_to_column()) %>% 
    dplyr::rename(factor=rowname, 
                  relVar.total=estimate, 
                  lower.total=`2.5%`, 
                  upper.total=`97.5%`)
  return(res)
}

doublecheckFitConvergence <- function(dnms, 
                                      withBatch=TRUE) {
  # For some model fits, a convergence warning might be thrown sounding like this:
  #
  #         In checkConv(attr(opt, "derivs"), opt$par, ctrl = control$checkConv,  :
  #           Model failed to converge with max|grad| = 0.00333311 (tol = 0.002, component 1)
  #
  # This can in principle be overcome by chosing the alternative optimzation option 'bobyqa' (https://stackoverflow.com/questions/21344555/convergence-error-for-development-version-of-lme4).
  # Nevertheless, the estimates obtained by both the standard optimization and the 'bobyqa' optimization are exremely close to each other. In order to not reimplement the functionality of the 'rptr'-package with the bobyqa-option, we here assert that both estimates are close to identical.
  if(withBatch) {
    suppressWarnings(
      opti_std <- 
      lme4::lmer(n ~ fathersAgeAtConceptionInYears + mothersAgeAtConceptionInYears + (1|familyNr) + (1|batch),
             data = dnms) %>% 
      broom.mixed::tidy()
    )
    opti_bob <- 
      lme4::lmer(n ~ fathersAgeAtConceptionInYears + mothersAgeAtConceptionInYears + (1|familyNr) + (1|batch),
             data = dnms,
             control = lme4::lmerControl(optimizer="bobyqa")) %>% 
      broom.mixed::tidy()
  } else {
    suppressWarnings(
      opti_std <- 
      lme4::lmer(n ~ fathersAgeAtConceptionInYears + mothersAgeAtConceptionInYears + (1|familyNr),
             data = dnms) %>% 
      broom.mixed::tidy()
    )
    opti_bob <- 
      lme4::lmer(n ~ fathersAgeAtConceptionInYears + mothersAgeAtConceptionInYears + (1|familyNr),
             data = dnms,
             control = lme4::lmerControl(optimizer="bobyqa")) %>% 
      broom.mixed::tidy()
  }
  stopifnot(
      near(opti_std$estimate, opti_bob$estimate, tol=1e-3)
    )
}
```

Functions for augmentative statistics
```{r}
familyStats <- function(inova1, allGenome) {
  familys <- 
    filterMuts(inova1, allGenome) %>% 
    group_by(familyNr) %>% 
    summarise(no.offspring = length(unique(SampleID))) %>% 
    filter(no.offspring>=2) 
  no.muts <- 
    filterMuts(inova1, allGenome) %>% 
    filter(familyNr %in% familys$familyNr) %>% 
    nrow()
  no.familys <- 
    familys %>% 
    nrow()
  no.offspring <- 
    familys %>%
    with(sum(no.offspring))
  tibble(no.familys=no.familys, no.muts.pp=no.muts/no.offspring)
}
familyStatsFrTbl <- function(autism1, tblHdr = "unique_snv") {
  no.offspring <- sum(autism1[, tblHdr]>0)
  no.familys <- autism1$familyNr %>% unique() %>% length()
  no.muts <- sum(autism1[, tblHdr])
  tibble(no.familys=no.familys, no.muts.pp=no.muts/no.offspring)
}

```

Doing the calculations
```{r}
set.seed(4321)

allEstimates <- 
  bind_rows(
    filterAndCalculate(inova1, uniqGenome, withBatch = TRUE) %>%
      mutate(set = "inova1"),
    filterAndCalculate(inova2, uniqGenome, withBatch = TRUE) %>%
      mutate(set = "inova2"),
    filterAndCalculate(sasani %>% dplyr::select(-end), uniqGenome) %>%
      mutate(set = "sasani"),
    calcRelativeVarCorsWithCi(
      autism1 %>% 
        mutate(n = unique_snv), 
      withBatch = TRUE) %>%
      mutate(set = "autism1"),
    calcRelativeVarCorsWithCi(
      autism2 %>% 
        mutate(n = snvs_unique), 
      withBatch = TRUE) %>%
      mutate(set = "autism2")
  ) 
# When running the above command, the following warnings and messages might arise:
# 
# 1. Message "boundary (singular) fit: see ?isSingular"
#    This message is lauched because it is difficult to estimate the batch effect for the inova2 data given the experimental distribution of samples into batches and the estimate is close to zero. See also here: https://bbolker.github.io/mixedmodels-misc/glmmFAQ.html#singular-models-random-effect-variances-estimated-as-zero-or-correlations-estimated-as---1
#
# 2. Warning "In rptGaussian(formula, grname, data, CI, nboot, npermut, parallel,  : 1 rows containing missing values were removed"
#    This warning is lauched because of batch information for one offspring of the three-offspring family 102-01507 in the inova2 dataset is missing. The estimation of the batch effect is therefore based on the remaining data points.
#
# 3. Several warnings like "In checkConv(attr(opt, "derivs"), opt$par, ctrl = control$checkConv,  :  Model failed to converge with max|grad| = 0.00302837 (tol = 0.002, component 1)"
#    These warnings indicate difficulties of the standard fitting algorithm to accurately estimate the parameters. Following the recommendations for dealing with these warnings, we control that the parameter estimations are close to equal to the results obtained by another estimator (without warnings) in the function 'doublecheckFitConvergence' defined above. See also  https://bbolker.github.io/mixedmodels-misc/glmmFAQ.html#troubleshooting



familyEstimates <-
  allEstimates %>%
  filter(factor == "familyNr") %>%
  mutate(set = rep(c("inova1", "inova2", "sasani", "autism1", "autism2")),
         includesBatch = c(TRUE, TRUE, FALSE, TRUE, TRUE)) %>%
  mutate(region = c(rep("unique", 5))) %>%
  bind_cols(
    bind_rows(
    familyStats(inova1, uniqGenome),
    familyStats(inova2, uniqGenome),
    familyStats(sasani %>% dplyr::select(-end), uniqGenome),
    familyStatsFrTbl(autism1, "unique_snv"),
    familyStatsFrTbl(autism2, "snvs_unique"))
  )

allEstimates %>%
  knitr::kable()
```

Vizualisation
```{r}
ggplot(familyEstimates %>%
         mutate(set = fct_inorder(set)),
       aes(x=set, 
           y=relVar.total,
           ymin=lower.total,
           ymax=upper.total, 
           alpha=no.familys)) + 
  geom_pointrange() + 
  geom_point() + 
  geom_errorbar() +
  geom_hline(yintercept = 0) + 
  facet_wrap(~region, ncol=1) +
  labs(x=NULL, y="Family variance component") + 
  ggpubr::theme_pubr()

allEstimates %>%
    mutate(set = fct_inorder(set),
           factor = fct_relevel(factor, c("Residual", "batch", "familyNr", "Fixed"))) %>%
  ggplot(aes(y=relVar.total, fill=factor, x=set)) + 
    geom_bar(stat="identity") + 
    ggpubr::theme_pubr()
```


Weighted results
```{r}
wMeanEstimate <- weighted.mean(
  x=familyEstimates$relVar.total,
  w=familyEstimates$no.familys
)

wSdEstimate <- Hmisc::wtd.var(
  x=familyEstimates$relVar.total,
  w=familyEstimates$no.familys
) %>% sqrt()

wMeanEstimate
wMeanEstimate + wSdEstimate * c(-1.96,1.96)
```



Poisson model comparisons
=========================


First, on all substitutions
```{r}
set.seed(4321)

ageMdlFrameList <- 
  list(
    inova1 = inova1 %>% 
      filterMuts(uniqGenome) %>%
      group_by(SampleID, fathersAgeAtConceptionInYears, mothersAgeAtConceptionInYears) %>% 
      dplyr::count() %>% 
      ungroup() %>% 
      complete(nesting(SampleID, 
                       fathersAgeAtConceptionInYears, 
                       mothersAgeAtConceptionInYears),
               fill = list(n=0)),
    inova2 = inova2 %>% 
      filterMuts(uniqGenome) %>%
      group_by(SampleID, fathersAgeAtConceptionInYears, mothersAgeAtConceptionInYears) %>% 
      dplyr::count() %>% 
      ungroup() %>% 
      complete(nesting(SampleID, 
                       fathersAgeAtConceptionInYears, 
                       mothersAgeAtConceptionInYears), 
               fill = list(n=0)),
    sasani = sasani %>% 
      dplyr::select(-end) %>% 
      filterMuts(uniqGenome) %>%
      group_by(SampleID, fathersAgeAtConceptionInYears, mothersAgeAtConceptionInYears) %>% 
      dplyr::count() %>% 
      ungroup() %>% 
      complete(nesting(SampleID, 
                       fathersAgeAtConceptionInYears, 
                       mothersAgeAtConceptionInYears), 
               fill = list(n=0))
  )


mdls <- 
  foreach(ageMdlFrame = ageMdlFrameList) %do% {
    glm(n ~ fathersAgeAtConceptionInYears + mothersAgeAtConceptionInYears, 
            data=ageMdlFrame, 
            family = poisson)
  }

countFrameList <- 
  foreach(n = 1:length(ageMdlFrameList)) %do% {
    ageMdlFrame <- ageMdlFrameList[[n]]
    mdlList <- mdls[[n]]
    poisDens <- 
      tibble(
        no.muts = 0:75,
        dens = getDensitySum(mdlFrame = ageMdlFrame, mdl = mdlList))
    countFrame <- 
      ageMdlFrame %>%
      dplyr::count(n) %>%
      ungroup() %>% 
      dplyr::rename(no.muts = n,
                    observed = nn) %>%
      full_join(poisDens %>%
                  mutate(predicted = round(dens))%>% #rounding the estimates
                  dplyr::select(-dens), 
                by = c("no.muts"))
    countFrame <- 
      countFrame %>%
      mutate(observed = if_else(is.na(observed), 0L, observed))
    return(countFrame)
  }

poisModPlots <- foreach(countFrame = countFrameList) %do% {
  ggplot(countFrame,
       aes(x=no.muts)) +
  geom_line(aes(y=predicted), col="red") + 
  geom_point(aes(y=observed)) + 
  xlim(0, 75) + 
  labs(y = "Number of offspring", x="Number of DNMs") + 
  ggpubr::theme_pubr()
}
cowplot::plot_grid(plotlist = poisModPlots,
                   nrow = 1,
                   labels = c("a", "b", "c"))

comparisons <- 
  foreach(countFrame = countFrameList,
          .combine = bind_rows) %do% {
            countFrame %>% 
              tidyr::gather("set", "cases", 2:3) %>%
              group_modify(~data.frame(wilcox.p=coin::wilcox_test(cases~as.factor(set), data=.x) %>% 
                                         coin::pvalue() %>% 
                                         as.numeric(),                                      #test for difference in median
                                       leveneTest(cases~factor(set), data=.x)$P[1],         #         heterogeneity of variance
                                       fligner.p=fligner.test(cases~set, data=.x)$p.value,  #         heterogeneity of variance
                                       ansari.p = coin::ansari_test(cases~as.factor(set), data=.x) %>% 
                                         coin::pvalue() %>% 
                                         as.numeric(),                                      #         difference in scale parameters
                                       mood.p   =   mood.test(cases~set, data=.x)$p.value,  #         difference in scale parameters
                                       f.p = var.test(no.muts~set, data=.x)$p.value))       #         compare the variances (parametric) 
          }

comparisons %>%
  knitr::kable()


```


Second, specifcally per substittution.

Get the data in shape
```{r}
ageMdlFrameList <- 
  list(
    inova1 = inova1 %>% 
      filterMuts(uniqGenome) %>%
      mutate(substitution = if_else(is.CpG, paste(substitution, "CpG"), as.character(substitution))) %>% 
      group_by(SampleID, fathersAgeAtConceptionInYears, mothersAgeAtConceptionInYears, substitution) %>% 
      dplyr::count() %>% 
      ungroup() %>% 
      filter(!is.na(substitution)) %>%
      complete(nesting(SampleID, 
                       fathersAgeAtConceptionInYears, 
                       mothersAgeAtConceptionInYears), 
               substitution, 
               fill = list(n=0)),
    inova2 = inova2 %>% 
      filterMuts(uniqGenome) %>%
      mutate(substitution = if_else(substitution == "C - T" & substr(surrounding,3,3) == "G", paste(substitution, "CpG"), as.character(substitution))) %>% 
      group_by(SampleID, fathersAgeAtConceptionInYears, mothersAgeAtConceptionInYears, substitution) %>% 
      dplyr::count() %>% 
      ungroup() %>% 
      filter(!is.na(substitution)) %>%
      complete(nesting(SampleID, 
                       fathersAgeAtConceptionInYears, 
                       mothersAgeAtConceptionInYears), 
               substitution, 
               fill = list(n=0)),
    sasani = sasani %>% 
      dplyr::select(-end) %>% 
      filterMuts(uniqGenome) %>%
      mutate(substitution = mut) %>% 
      group_by(SampleID, fathersAgeAtConceptionInYears, mothersAgeAtConceptionInYears, substitution) %>% 
      dplyr::count() %>% 
      ungroup() %>% 
      filter(!is.na(substitution)) %>%
      complete(nesting(SampleID, 
                       fathersAgeAtConceptionInYears, 
                       mothersAgeAtConceptionInYears), 
               substitution, 
               fill = list(n=0))
  )

mdls <- 
  foreach(ageMdlFrame = ageMdlFrameList) %do% {
    mdlList<- 
      foreach(subs = unique(ageMdlFrame$substitution)) %do% {
        glm(n ~ fathersAgeAtConceptionInYears + mothersAgeAtConceptionInYears, 
            data=ageMdlFrame, 
            family = poisson,
            subset = substitution == subs)
      }
    names(mdlList) <- unique(ageMdlFrame$substitution)
    return(mdlList)
  }

```


calculate
```{r}
set.seed(4321)

countFrameList <- 
  foreach(n = 1:length(ageMdlFrameList)) %do% {
    ageMdlFrame <- ageMdlFrameList[[n]]
    mdlList <- mdls[[n]]
    poisDens <- 
      foreach(subs = unique(ageMdlFrame$substitution),
              .combine = bind_rows) %do% {
                tibble(
                  no.muts = 0:75,
                  dens = getDensitySums(subs, mdlFrame = ageMdlFrame, mdls = mdlList), 
                  substitution = subs)
              }
    countFrame <- 
      ageMdlFrame %>%
      group_by(substitution) %>%
      dplyr::count(n) %>%
      ungroup() %>% 
      dplyr::rename(no.muts = n,
                    observed = nn) %>%
      full_join(poisDens %>%
                  mutate(predicted = round(dens))%>% #rounding the estimates
                  dplyr::select(-dens), 
                by = c("substitution", "no.muts"))
    countFrame <- 
      countFrame %>%
      mutate(observed = if_else(is.na(observed), 0L, observed))
    return(countFrame)
  }
```


visualize
```{r}
foreach(countFrame = countFrameList) %do% {
  ggplot(countFrame,
       aes(x=no.muts)) +
  geom_line(aes(y=predicted), col="red") + 
  geom_point(aes(y=observed)) + 
  facet_wrap(~substitution) + 
  xlim(0, 40) + 
  labs(y = "Number of DNMs", x = "Number of Offspring") + 
    ggpubr::theme_pubr()
}
```


```{r}
comparisons <- 
  foreach(countFrame = countFrameList,
          .combine = bind_rows) %do% {
            countFrame %>% 
              tidyr::gather("set", "cases", 3:4) %>%
              group_by(substitution) %>% 
              group_modify(~data.frame(wilcox.p=coin::wilcox_test(cases~as.factor(set), data=.x) %>% 
                                         coin::pvalue() %>% 
                                         as.numeric(),                                      #test for difference in median
                                       leveneTest(cases~factor(set), data=.x)$P[1],         #         heterogeneity of variance
                                       fligner.p=fligner.test(cases~set, data=.x)$p.value,  #         heterogeneity of variance
                                       ansari.p = coin::ansari_test(cases~as.factor(set), data=.x) %>% 
                                         coin::pvalue() %>% 
                                         as.numeric(),                                      #         difference in scale parameters
                                       mood.p   =   mood.test(cases~set, data=.x)$p.value,  #         difference in scale parameters
                                       f.p = var.test(no.muts~set, data=.x)$p.value))       #         compare the variances (parametric) 
          }

comparisons %>%
  knitr::kable()
```

